feature-libs/asm/root/interceptors/user-id.interceptor.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * SPDX-FileCopyrightText: 2022 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {
+  HttpEvent,
+  HttpHandler,
+  HttpInterceptor,
+  HttpRequest,
+} from '@angular/common/http';
+import { Inject, Injectable } from '@angular/core';
+import {
+  OCC_ASM_TOKEN,
+  OCC_USER_ID_CONSTANTS,
+  UserIdService,
+} from '@spartacus/core';
+import { Observable, of } from 'rxjs';
+import { concatMap, map } from 'rxjs/operators';
+
+/**
+ * Looks for a specific key in the HttpRequest's context (OCC_ASM_TOKEN) to decide when to
+ * configure a request with 'sap-commerce-cloud-user-id' header.
+ */
+@Injectable({ providedIn: 'root' })
+export class UserIdInterceptor implements HttpInterceptor {
```

Comment:
Could this logic become non ASM specific and make it's way in the core lib?  Let's discuss.

Reply:
That's an interesting discussion. It was moved to the ASM module based on this premise:
1. 'sap-commerce-cloud-user-id' only needs to be sent when the user needs to be explicit i.e. in customer emulation case.
2. If it's only for customer emulation, then we can bolster the code's performance, if ever so slightly, by removing this interceptor from the chain, and by needing to load the code in the first place.

Is there an incentive to move it to core?

Reply:
In OCC/backend, is this header used exclusively for ASM?

Reply:
The header essentially overrides any session context that may have been used. For example, passing in the emulated customer ID will always override the user agent's authorization with the emulated user's. It's a very straightforward, if blunt, tool.

I can only imagine this would be used for ASM _for now_. I can't imagine any other scenario, at least given the current use of SPA.

Reply:
I'm ok to leave the interceptor in asm.  I agree that for the foreseeable future, the purpose of the interceptor is to enable customer emulation for endpoints that don't have other ways to pass the user id parameter.